### PR TITLE
Improve CLI and API reading entry points

### LINKED ISSUE #21: CLI for summarizing and converting between data formats
`recOrder` provides CLI for summarizing ND data without loading it, and converting from OME-TIFF to OME-zarr format. These CLI are more broadly useful and therefore should be moved to `iohub`.

The consensus was to use the `click` library to implement CLI.

### LINKED ISSUE #90: NGFF version is not found by `imread`
Running:
```
python single_fov_ome_zarr.py
iohub info ~/ome.zarr
```
yields:
```
Traceback (most recent call last):                                                                                    |103 # this array is about 10 GB, each time point is about 100 MB                    |
  File "/opt/anaconda3/envs/iohub-dev/bin/iohub", line 8, in <module>                                                 |104 # it may not be actually larger than RAM but enough for demo                    |
    sys.exit(cli())                                                                                                   |105 # monitor the memory usage of python when the following runs                    |
  File "/opt/anaconda3/envs/iohub-dev/lib/python3.9/site-packages/click/core.py", line 1130, in __call__              |106 # and it should take significantly less than 10 GB                              |
    return self.main(*args, **kwargs)                                                                                 |107 large_tczyx = da.ones((300, 2, 25, 1024, 1024), dtype=np.uint16)                |
  File "/opt/anaconda3/envs/iohub-dev/lib/python3.9/site-packages/click/core.py", line 1055, in main                  |108                                                                                 |
    rv = self.invoke(ctx)                                                                                             |109 if store_path:                                                                  |
  File "/opt/anaconda3/envs/iohub-dev/lib/python3.9/site-packages/click/core.py", line 1657, in invoke                |110     with open_ome_zarr(                                                         |
    return _process_result(sub_ctx.command.invoke(sub_ctx))                                                           |111         store_path, layout="fov", mode="w-", channel_names=["DAPI", "GFP"]      |
  File "/opt/anaconda3/envs/iohub-dev/lib/python3.9/site-packages/click/core.py", line 1404, in invoke                |112     ) as dataset:                                                               |
    return ctx.invoke(self.callback, **ctx.params)                                                                    |113         img = dataset.create_zeros(                                             |
  File "/opt/anaconda3/envs/iohub-dev/lib/python3.9/site-packages/click/core.py", line 760, in invoke                 |114             name="0",                                                           |
    return __callback(*args, **kwargs)                                                                                |115             shape=large_tczyx.shape,                                            |
  File "/Users/talon.chandler/iohub/iohub/cli/cli.py", line 22, in info                                               |116             dtype=large_tczyx.dtype,                                            |
    reader = imread(file)                                                                                             |117             chunks=(1, 1, 1, 1024, 1024),  # chunk by XY planes                 |
  File "/Users/talon.chandler/iohub/iohub/reader.py", line 177, in imread                                             |118         )                                                                       |
    return ZarrReader(path, version=ngff_version)                                                                     |119         for t, not_so_large_czyx in tqdm(enumerate(large_tczyx)):               |
  File "/Users/talon.chandler/iohub/iohub/zarrfile.py", line 50, in __init__                                          |120             img[t] = not_so_large_czyx.compute()                                |
    raise ValueError(f"Invalid NGFF version: {version}")                                                              |121         dataset.print_tree()                                                    |
ValueError: Invalid NGFF version: True
```

This might be a shallow bug. It seems that NGFF stores the version at `['omero']['version']` instead of `['plate']['version']`. See https://github.com/czbiohub/iohub/blob/26855833ddae316e1a5a1561e717aa70b07e886e/iohub/reader.py#L42

### PR DESCRIPTION
**Pull request by ziw-liu**
Updated after offline [meeting 2023-03-20](https://github.com/czbiohub/iohub/wiki/2023-03-20-meeting-notes/cb483b2bd64e23d55478590157abdf6c20778ffb):

**Behavior changes:**
- Rename `imread()` to `read_micromanager()` to be more specific.
- <del>`imread()` now uses `open_ome_zarr()` for NGFF v0.4 datasets, which means that the returned object can have different parent types (`ReaderBase` or `NGFFNode`) and methods.</del>
- If <del>`imread()`</del>`read_micromanager()` is used to open NGFF v0.4 datasets, a warning is shown to recommend switching to `open_ome_zarr()`.
- `iohub info -v /path/to/data` now shows python code example to read the data (`imread()` or `open_ome_zarr()`)

**Fixes:**
- Now <del>`imread()`</del>`read_micromanager()` and `open_ome_zarr()` are both importable from the top-level `iohub` namespace (I put code in the wrong file in #71)
- The `iohub info` CLI command now works for all data formats

Fixes #90 and resolves #21.
Partial implementation of #40. Won't close it yet but probably can defer more work on that to the next version.

### CODE REVIEW

**Comment by ziw-liu** [comment]
Still fixing tests...

**Review by talonchandler — COMMENTED** [review]
Very nice! I don't have any major concerns. 

I've gone back and forth on if `imread` should return different objects. I still have a soft vote for "no"---I slightly prefer `imread` to return one object type and `open_ome_zarr` to return the other object type until the APIs are more similar---but I won't block on this point. If we stick with `imread` returning two types I think the warning should be more explicit (see my comment).

File: iohub/reader.py
Comment (ziw-liu) [reply iohub/reader.py:270]: How about `=== Summary ===` then? I actually just wanted a separation line here so `=======` also works.

Comment (talonchandler) [reply iohub/reader.py:270]: I like it a lot!

File: iohub/reader.py
Comment (ziw-liu) [reply iohub/reader.py:148]: As per Python [docs](https://docs.python.org/3/library/exceptions.html):

> In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.

My understanding is that this is used in abstract classes (e.g. `ReaderBase`) and not buisiness logic code. But I agree that `FileNotFoundError` can be confusing, maybe a more generic `RuntimeError`?

**Comment by ziw-liu** [comment]
OME-TIFF reader tests start failing after changing something completely unrelated. Also can't reproduce locally...

File: iohub/reader.py
Comment (ziw-liu) [reply iohub/reader.py]: Done!

File: iohub/reader.py
Comment (talonchandler) [reply iohub/reader.py]: LGTM!

File: iohub/reader.py
Comment (talonchandler) [inline iohub/reader.py:217]: Nice!

**Review by talonchandler — APPROVED** [review]
LGTM! I'm approving. Maybe consider waiting until after Monday's `iohub` meeting to merge?

File: iohub/reader.py
Comment (talonchandler) [inline iohub/reader.py:213]: imread -> read_micromanager here in these two lines

**Review by talonchandler — APPROVED** [review]
I think `read_micromanager` is an appropriate name. The `read_micromanager` can read v0.1 ome.zarrs as well, but all of these datasets originated from micromanager datasets so it makes sense to me. 

`read_micromanager` and `open_ome_zarr` play similar roles (even though the objects they are different and have different APIs). Maybe consider `open_micromanager` (or alternatively `read_ome_zarr`)? I won't block if there's a good reason for this difference.

**Comment by ziw-liu** [comment]
> Maybe consider `open_micromanager` (or alternatively `read_ome_zarr`)? I won't block if there's a good reason for this difference.

The main difference is that the readers only read, while `open_ome_zarr()` implements a file mode syntax mimicking the built-in `open()` that can create/write/delete data.

File: iohub/reader.py
Comment (ziw-liu) [reply iohub/reader.py:213]: Thanks for the catch, fixing.

Comment (ziw-liu) [reply iohub/reader.py:213]: Actually this warning message is more outdated than just the name...

**Comment by talonchandler** [comment]
Thanks @ziw-liu. I think everything here makes sense, and it's ready to merge from my perspective. 

@mattersoflight @AhmetCanSolak, can I request your comment so that this PR is no longer a bottleneck? After this merges I am planning to tag the latest commit and start depending on it elsewhere via github (no PyPI release). See #55 for more context on my immediate next steps.

**Review by AhmetCanSolak — APPROVED** [review]
thank you @ziw-liu for addressing all the raised concerns regarding these API updates.  Also, additional warnings for promoting `iohub.open_ome_zarr()` is great. I am happy with the current state of this PR and I would this should be merged soon, since it is a blocker to @talonchandler .

**Review by mattersoflight — APPROVED** [review]
The new API reads well.  Thanks @ziw-liu! 

I am fine with `read_micromanager` entry point giving access to the earlier zarr stores.

I tried `iohub info` and am appending the transcript.

There are a couple of quick-to-fix issues.
* `iohub info` should show a help message when it is called without an argument. At this point, it just returns.
* The help message should be updated to note recognized file formats (Micro-Manager TIFF, ND TIff, NGFF v0.4, etc.)
* If the data is not of a format recognized by `iohub info`, the error message should point that out.

---------------- 
```
(mantis) ➜  iohub git:(ngff-info) iohub info           
(mantis) ➜  iohub git:(ngff-info) iohub info --help
Usage: iohub info [OPTIONS] [FILES]...

  View basic metadata of a list of FILES

Options:
  -h, --help     Show this message and exit.
  -v, --verbose  Show usage guide to open dataset in Python and full tree for
                 HCS Plates in OME-Zarr
(mantis) ➜  iohub git:(ngff-info) iohub info ~/hcs.zarr
Reading file:    /Users/shalin.mehta/hcs.zarr

=== Summary ===
Format:          omezarr v0.4
Axes:            T (time); C (channel); Z (space); Y (space); X (space); 
Channel names:   ['DAPI', 'GFP', 'Brightfield', 'Segmentation']
Row names:       ['A', 'H', 'Control']
Column names:    ['1', '12', 'Blank']
Wells:           4
(mantis) ➜  iohub git:(ngff-info) iohub info ~/hcs.zarr/A/1/0
Reading file:    /Users/shalin.mehta/hcs.zarr/A/1/0
Zarr hierarchy:
/
 └── 0 (5, 4, 2, 32, 32) uint16

=== Summary ===
Format:          omezarr v0.4
Axes:            T (time); C (channel); Z (space); Y (space); X (space); 
Channel names:   ['DAPI', 'GFP', 'Brightfield', 'Segmentation']
(mantis) ➜  iohub git:(ngff-info) iohub info ~/hcs.zarr/A/1/ 
Reading file:    /Users/shalin.mehta/hcs.zarr/A/1/
Traceback (most recent call last):
  File "/Users/shalin.mehta/miniconda3/envs/mantis/bin/iohub", line 8, in <module>
    sys.exit(cli())
  File "/Users/shalin.mehta/miniconda3/envs/mantis/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/shalin.mehta/miniconda3/envs/mantis/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/shalin.mehta/miniconda3/envs/mantis/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/shalin.mehta/miniconda3/envs/mantis/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/shalin.mehta/miniconda3/envs/mantis/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/shalin.mehta/Documents/code/iohub/iohub/cli/cli.py", line 31, in info
    print_info(file, verbose=verbose)
  File "/Users/shalin.mehta/Documents/code/iohub/iohub/reader.py", line 250, in print_info
    reader = read_micromanager(path, data_type=fmt)
  File "/Users/shalin.mehta/Documents/code/iohub/iohub/reader.py", line 218, in read_micromanager
    raise ValueError(f"NGFF version {extra_info} is not supported.")
ValueError: NGFF version unknown is not supported.
```

**Comment by mattersoflight** [comment]
@ziw-liu Tested - looks good!

**Comment by ziw-liu** [comment]
```text
---------- coverage: platform darwin, python 3.9.16-final-0 ----------
Name                      Stmts   Miss  Cover
---------------------------------------------
iohub/__init__.py             4      0   100%
iohub/_version.py             2      0   100%
iohub/cli/__init__.py         0      0   100%
iohub/cli/cli.py             27      0   100%
iohub/convert.py            156     24    85%
iohub/lf_utils.py            12      1    92%
iohub/multipagetiff.py      156     25    84%
iohub/ndtiff.py              69      3    96%
iohub/ngff.py               467     65    86%
iohub/ngff_meta.py          181     20    89%
iohub/reader.py             145     15    90%
iohub/reader_base.py         41      4    90%
iohub/singlepagetiff.py     161     21    87%
iohub/upti.py                86     69    20%
iohub/zarrfile.py           137     29    79%
---------------------------------------------
TOTAL                      1644    276    83%
```

**Pull request merged**
